hazelcast/src/main/java/com/hazelcast/internal/util/AmbigiousInstantiationException.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util;
+
+import com.hazelcast.core.HazelcastException;
+
+/**
+ * Indicates there are multiple constructors matching given parameters.
+ *
+ */
+final class AmbigiousInstantiationException extends HazelcastException {
```

Comment:
I am not sure if a new exception is the right choice. however it seems like the best thing to be thrown from `InstantiationUtils`

Reply:
Why not make use of IllegalArgument or IllegalStateException and provide a proper exception message? These messages are not recoverable; we don't expect people to catch them.

Reply:
My thinking was catching `AmbigiousInstantiationException` can indeed be useful, even in our own code. Example: I was considering catching it even in the single case where this is used - and wrap it inside `ConfigurationException`. 

Maybe `AmbigiousInstantiationException` could indeed extends the `IllegalArgumentException`.

Reply:
If we want to catch it, I'm fine having a custom class.
If we don't want to catch it, it just leads to more noise imho.